File: helpers/convert-manifest.py
import json
import re
import requests
from collections import defaultdict

filtered_scripts = ['scripts/common.js', 'scripts/global.js', 'scripts/store/invalidate_cache.js']
filtered_css = ['styles/global.css', 'styles/store.css', 'styles/community.css']

def url_to_regex(url):
    return re.escape(url).replace(r'\*', '.*')

def convert_to_js(manifest_file):
    response = requests.get(manifest_file)
    data = json.loads(response.text)

    content_scripts = data['content_scripts']
    combined_matches = defaultdict(lambda: {'js': [], 'css': []})

    for script in content_scripts[1:]:
        matches = script['matches']
        js_files = script.get('js', [])
        css_files = script.get('css', [])

        # Convert matches to regex using custom function
        combined_matches_str = ' || '.join(['href.match("{}")'.format(url_to_regex(m)) for m in matches])
        combined_matches[combined_matches_str]['js'].extend(js_files)
        combined_matches[combined_matches_str]['css'].extend(css_files)

    for match, files in combined_matches.items():
        js_files = files['js']
        css_files = files['css']
        
        unique_js_files = list(dict.fromkeys(js_files))
        unique_css_files = list(dict.fromkeys(css_files))
        
        filtered_js_files = [js_file for js_file in unique_js_files if js_file not in filtered_scripts]
        filtered_css_files = [css_file for css_file in unique_css_files if css_file not in filtered_css]
        
        if len(filtered_js_files) == 0 and len(filtered_css_files) == 0:
            continue

        print('if ({}) {{'.format(match))
        for js_file in filtered_js_files:
            print('    scripts.push("{}");'.format(js_file))
        for css_file in filtered_css_files:
            print('    scripts.push("{}");'.format(css_file))
        print('}\n')

if __name__ == "__main__":
    convert_to_js('https://cdn.jsdelivr.net/gh/SteamDatabase/BrowserExtension@4.10/manifest.json')

File: helpers/convert-manifest.ts
/* eslint-disable @typescript-eslint/no-loop-func */
import clipboard from 'clipboardy';
import escapeStringRegexp from 'escape-string-regexp';
import * as fs from 'fs';

interface ContentScript {
  matches: string[];
  exclude_matches?: string[];
  js?: string[];
  css?: string[];
}

interface ManifestData {
  content_scripts: ContentScript[];
}

function urlToRegex(url: string): string {
  return `^${escapeStringRegexp(url).replace(/\\\*/g, '.*').replace(/\//g, '\\/')}$`;
}

function convertToJs(): void {
  fs.readFile('AugmentedSteam/dist/prod.chrome/manifest.json', 'utf8', (err, data) => {
    if (err) {
      console.error('Error reading the manifest file:', err);

      return;
    }

    const manifestData = JSON.parse(data) as ManifestData;
    const contentScripts = manifestData.content_scripts;
    const combinedMatches: Record<string, { js: string[]; css: string[]; }> = {};

    let output = '';

    for (const script of contentScripts) {
      const matches = script.matches;
      const excludes = script.exclude_matches;
      const jsFiles = script.js ?? [];
      const cssFiles = script.css ?? [];

      let combinedMatchesStr = matches.map(m => `href.match(/${urlToRegex(m)}/)`).join(' || ');
      combinedMatchesStr = `(${combinedMatchesStr})`;
      // Add the exclude_matches to the combinedMatchesStr
      if (excludes) {
        combinedMatchesStr += ` && !(${excludes.map(m => `href.match(/${urlToRegex(m)}/)`).join(' || ')})`;
      }
      if (!combinedMatches[combinedMatchesStr]) {
        combinedMatches[combinedMatchesStr] = { js: [], css: [] };
      }
      combinedMatches[combinedMatchesStr]?.js.push(...jsFiles);
      combinedMatches[combinedMatchesStr]?.css.push(...cssFiles);
    }

    for (const [match, { js, css }] of Object.entries(combinedMatches)) {
      const uniqueJsFiles = Array.from(new Set(js));
      const uniqueCssFiles = Array.from(new Set(css));

      if (uniqueJsFiles.length === 0 && uniqueCssFiles.length === 0) {
        continue;
      }

      output += `if (${match}) {\n`;
      uniqueJsFiles.forEach((jsFile) => {
        output += `    scripts.push('${jsFile}');\n`;
      });
      uniqueCssFiles.forEach((cssFile) => {
        output += `    scripts.push('${cssFile}');\n`;
      });
      output += '}\n\n';
    }

    output = output.trimEnd();

    // Copy text to clipboard
    clipboard.writeSync(output);

    console.log('Copied to clipboard.');
  });
}

convertToJs();
